url/url.go

```diff
@@ -0,0 +1,331 @@
+		}
+	})
+
+	// hash
+	defineURLAccessorProp(r, p, "hash", func(u *url.URL) interface{} {
+		if u.Fragment != "" {
+			return "#" + u.EscapedFragment()
+		}
+		return ""
+	}, func(u *url.URL, arg goja.Value) {
+		h := arg.String()
+		if len(h) > 0 && h[0] == '#' {
+			h = h[1:]
+		}
+		u.Fragment = h
+	})
+
+	// hostname
+	defineURLAccessorProp(r, p, "hostname", func(u *url.URL) interface{} {
+		return strings.Split(u.Host, ":")[0]
+	}, func(u *url.URL, arg goja.Value) {
+		h := arg.String()
+		if strings.IndexByte(h, ':') >= 0 {
+			return
+		}
+		if _, err := url.ParseRequestURI(u.Scheme + "://" + h); err == nil {
+			if port := u.Port(); port != "" {
+				u.Host = h + ":" + port
+			} else {
+				u.Host = h
+			}
+			fixURL(r, u)
+		}
+	})
+
+	// href
+	defineURLAccessorProp(r, p, "href", func(u *url.URL) interface{} {
+		return u.String()
+	}, func(u *url.URL, arg goja.Value) {
+		url := parseURL(r, arg.String(), true)
+		*u = *url
+	})
+
+	// pathname
+	defineURLAccessorProp(r, p, "pathname", func(u *url.URL) interface{} {
+		return u.EscapedPath()
+	}, func(u *url.URL, arg goja.Value) {
+		p := arg.String()
+		if _, err := url.Parse(p); err == nil {
+			switch u.Scheme {
+			case "https", "http", "ftp", "ws", "wss":
+				if !strings.HasPrefix(p, "/") {
+					p = "/" + p
+				}
+			}
+			u.Path = p
+		}
+	})
+
+	// origin
+	defineURLAccessorProp(r, p, "origin", func(u *url.URL) interface{} {
+		return u.Scheme + "://" + u.Hostname()
+	}, nil)
+
+	// password
+	defineURLAccessorProp(r, p, "password", func(u *url.URL) interface{} {
+		p, _ := u.User.Password()
+		return p
+	}, func(u *url.URL, arg goja.Value) {
+		user := u.User
+		u.User = url.UserPassword(user.Username(), arg.String())
+	})
+
+	// username
+	defineURLAccessorProp(r, p, "username", func(u *url.URL) interface{} {
+		return u.User.Username()
+	}, func(u *url.URL, arg goja.Value) {
+		p, has := u.User.Password()
+		if !has {
+			u.User = url.User(arg.String())
+		} else {
+			u.User = url.UserPassword(arg.String(), p)
+		}
+	})
+
+	// port
+	defineURLAccessorProp(r, p, "port", func(u *url.URL) interface{} {
+		return u.Port()
+	}, func(u *url.URL, arg goja.Value) {
+		setURLPort(u, arg)
+	})
+
+	// protocol
+	defineURLAccessorProp(r, p, "protocol", func(u *url.URL) interface{} {
+		return u.Scheme + ":"
+	}, func(u *url.URL, arg goja.Value) {
+		s := arg.String()
+		pos := strings.IndexByte(s, ':')
+		if pos >= 0 {
+			s = s[:pos]
+		}
+		s = strings.ToLower(s)
+		if isSpecialProtocol(u.Scheme) == isSpecialProtocol(s) {
+			if _, err := url.ParseRequestURI(s + "://" + u.Host); err == nil {
+				u.Scheme = s
+			}
+		}
+	})
+
+	// Search
+	defineURLAccessorProp(r, p, "search", func(u *url.URL) interface{} {
+		if u.RawQuery != "" {
+			return "?" + u.RawQuery
+		}
+		return ""
+	}, func(u *url.URL, arg goja.Value) {
+		u.RawQuery = arg.String()
+		fixRawQuery(u)
+	})
+
+	// search Params
+	defineURLAccessorProp(r, p, "searchParams", func(u *url.URL) interface{} {
```

Comment:
Most of the code in this file was simply moved from `module.go`. I've only added this new method as part of the URL implementation (Along with tests).